---
title: "The `SpatialData` class"
date: "`r format(Sys.Date(), '%B %d, %Y')`"
package: "`r BiocStyle::pkg_ver('SpatialData')`"
author:
- name: Constantin Ahlmann-Eltze
  affiliation: EMBL Genome Biology Unit, Heidelberg, Germany
- name: Helena L Crowell
  affiliation:
  - Department of Molecular Life Sciences, University of Zurich, Switzerland
  - SIB Swiss Institute of Bioinformatics, University of Zurich, Switzerland
- name: Tim Treis
  affiliation: Computational Health Center, Helmholtz Center Munich, Neuherberg, Germany
output: 
  BiocStyle::html_document:
    toc: true
    toc-depth: 2
    toc-float: true
# having some bug & this gets around it
knit: (function(input, ...) rmarkdown::render(input, output_dir=".")) 
vignette: >
  %\VignetteIndexEntry{"foo"}
  %\VignettePackage{foo}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
#bibliography: "`r file.path(system.file('extdata', package='muscat'), 'refs.bib')`"
abstract: > 
  <p> ...
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(cache=TRUE)
```

# Preamble

## Dependencies

```{r load-libs, message=FALSE, warning=FALSE}
library(Rarr)
library(EBImage)
library(ggplot2)
library(jsonlite)
library(patchwork)
library(SpatialData)
library(SingleCellExperiment)
```

## Introduction

# `SpatialData`

The `SpatialData` class contains 5 elements that are represented as follows:

- `images`: list of `ImageArray`s
- `labels`: list of `LabelArray`s
- `shapes`: list of `ShapeFrame`s
- `points`: list of Arrow `Dataset`s
- `table`: a `SingleCellExperiment`

```{r}
path <- system.file("extdata", "raccoon", package="SpatialData", mustWork=TRUE)
(spd <- readSpatialData(path))
```

## Accession

The following accessors are currently supported:

- `image/label/shape/pointNames` to retrieve available entities of the respective element.
- `images/labels/shapes/points` to retrieve a list of entities of the respective element.
- `image/label/shape/point` to retrieve a single entity of the respective element.
- `$` to directly access the entities of the respective element.

`SpatialData` objects behave like a list, i.e., entities of a all
elements can be accessed in various (equivalent) ways:

```{r results="hide"}
# these are all equivalent
i <- imageNames(spd)[1]
element(spd, "images", i)
images(spd)[[i]]
image(spd, i)
spd$images[[i]]
```

## Elements

The `ZarrArray` class is essentially an `Annotated` array-like object, that may
contain a dense `array` or any type of `Array` (e.g., `Sparse/DelayedArray`).
Derived here-from are the `ImageArray` and `LabelArray` classes
that represent single entities of `images` and `labels`, respectively.
These differ slightly in their associated metadata
and array properties, but share many functions.

```{r}
# extract 'ImageArray'
img <- image(spd, i)
# rename channels
channels(img) <- c("R", "G", "B")
img
```

Shapes are represented as `ShapeFrames`.
Currently, these support shapes of type _circle_ and _polygon_.
Under the hood, these are just `DataFrame`s
(with specialized methods and additional internal requirements):

```{r}
(s <- shape(spd))
DataFrame(s)
```

# Transformations

Only translation, scaling, and rotation of `Image/LabelArray`s are currently supported
via the following functions, each of which expects a \code{SpatialData} object (SD),
as well as transformation data `t` according to:

- `rotateElement`: scalar integer (angle).
- `scaleElement`: numeric vector with as many elements as SD has dimensions.
- `translateElement`: integer vector of length 2 (row/y and col/x translation).

Available coordinate systems can be retrieved via `coords()`:

- `type`: type of transformation (one of identity, rotate, scale, ...)
- `data`: transformation data (`NA` when `type = "identity"`)
- `output.name` the target coordinate system's identity

```{r}
coords(img)
```

## Translation

```{r}
# move the raccoon up'n' down
ps <- lapply(c(-300, 0, 300), \(t) 
    translateElement(img, c(t, 0)))
wrap_plots(lapply(ps, plotElement))
```

## Rotation

```{r}
# spin the raccoon (anti-clockwise)
ps <- lapply(c(0, 30, 60), \(t) 
    rotateElement(img, t))
wrap_plots(lapply(ps, plotElement))
```

## Scaling

```{r}
# squeeze the raccoon horizontally
ps <- lapply(c(1, 2/3, 1/2), \(t) 
    scaleElement(img, c(1, 1, t)))
wrap_plots(lapply(ps, plotElement))
```

# Visualization

## By element

Element-wise plotting is possible via `plotElement`,
which expects a single `SpatialData` element as input.
Images and shapes are rendered via `annotation_raster`;
for shapes (`geom_polygon`), additional graphical parameters
may be passed as dot (`...`) arguments.

```{r}
plotElement(image(spd)) +
plotElement(label(spd)) +
plotElement(shape(spd), col="blue", fill="cyan") 
```

## Layered

`plotSD` currently supports overlaying up to 3 elements 
(an image, label and shape, but not points yet).
Argument `coord` specifies the target coordinate system,
and will default the first available shared one if left unspecified.
Elements are internally aligned via `alignElements`,
which in turn calls `transformArray` on the input
`image` and `label` (type `Image/LabelArray`).
Depending on the underlying metadata, `transformElement`
uses `scale/rotate/translateElement` for transformation.

```{r}
# defaults to including all available elements
plotSD(spd,
    alpha.label=1/3,
    alpha.shape=1,
    fill.shape="lightgrey")
```

```{r}
path <- system.file("extdata", "blobs", package="SpatialData", mustWork=TRUE)
(spd <- readSpatialData(path))
plotElement(image(spd)) +
plotElement(label(spd)) +
plotElement(shape(spd)) 
plotSD(spd,
    alpha.label=1/3,
    alpha.shape=1,
    fill.shape="lightgrey")
```

# Query

## Aggregation

`aggregateImage` computed aggregated measurement values from `image`
according to `label` using `fun` to summarized measurements (default `mean`).
By default, the first available `image` and `label` are used.
In the output `SingleCellExperiment`, rows correspond
to channels and columns to unique labels (excluding 0); 
aggregated xy-coordinates are included in the `colData`.

```{r}
(sce <- aggregateImage(spd))
```

# Appendix

## Session info

```{r session-info}
sessionInfo()
```
